testah-junit/src/main/java/org/testah/framework/report/jira/AbstractJiraRemoteLinkBuilder.java

```diff
@@ -0,0 +1,263 @@
+        return icon2;
+    }
+
+    /**
+     * Gets last test plan dto used, the testplan the testcase methods will use to get source and run info.
+     *
+     * @return the last test plan dto used
+     */
+    protected TestPlanDto getLastTestPlanDtoUsed() {
+        if (null == lastTestPlanDtoUsed) {
+            throw new RuntimeException("Unable to use getLastTestPlanDtoUsed() unless you "
+                    + " call either getRemoteLinkForTestPlanResult or getRemoteLinkForTestPlanResultKnownProblem or "
+                    + "set lastTestPlanDtoUsed with setLastTestPlanDtoUsed");
+        }
+        return lastTestPlanDtoUsed;
+    }
+
+    /**
+     * Sets last test plan dto used. Can be used to set the testplan the testcase methods will use to get source and run info.
+     *
+     * @param lastTestPlanDtoUsed the last test plan dto used
+     */
+    public void setLastTestPlanDtoUsed(final TestPlanDto lastTestPlanDtoUsed) {
+        this.lastTestPlanDtoUsed = lastTestPlanDtoUsed;
+    }
+
+    /**
+     * Gets run link to use. Can be used to set the testplan the testcase methods will use to get source and run info
+     *
+     * @return the run link to use
+     */
+    protected String getRunLinkToUse() {
+        return validateUrl(getLastTestPlanDtoUsed().getRunInfo().getRunLocation(), "Use-Envir-Param=param_runLocation");
+    }
+
+    /**
+     * Gets source link to use.
+     *
+     * @return the source link to use
+     */
+    protected String getSourceLinkToUse() {
+        String sourceUrl = validateUrl(TS.params().getSourceUrl(), "IssueGetting_param_sourceUrl");
+        return sourceUrl + (sourceUrl.endsWith("/") ? "" : "/") + getSourceWithSlash(getLastTestPlanDtoUsed().getSource());
+    }
+
+    /**
+     * Gets source with slash. For use with source code mgmt, turn the classpath seperator from dot to slash.
+     *
+     * @param source the source is the absolute classpath to the testplan class
+     * @return the source with slash
+     */
+    protected String getSourceWithSlash(final String source) {
+        if (StringUtils.isEmpty(source)) {
+            return source;
+        }
+        return source.replace(".", "/");
+    }
+
+    /**
+     * Validate url string.  Checks that the string is not empty, if it is will put in a placeholder. Also ensures it starts with http
+     * Jira requires a valid link, and will fail if it is not.
+     *
+     * @param link     the link must be a valid link starting with http
+     * @param errorTip the error tip is added as a query parma to the fake link to help know what why fake link is used.
+     * @return the string the link or a value that is not empty using a fake link
+     */
+    protected String validateUrl(String link, final String errorTip) {
+        if (StringUtils.isEmpty(link)) {
+            return "http://noLinkFoundToUsePlease.com/?errorTip=" + errorTip;
+        }
+        if (!link.toLowerCase().startsWith("http")) {
```

Comment:
use https://commons.apache.org/proper/commons-lang/javadocs/api-3.6/src-html/org/apache/commons/lang3/StringUtils.html#line.7386 to simplify plus https://commons.apache.org/proper/commons-lang/javadocs/api-3.6/org/apache/commons/lang3/StringUtils.html#startsWith-java.lang.CharSequence-java.lang.CharSequence-